rudimentary performance monitoring

tomeichlersmith opened the pull request.

tomeichlersmith committed:
WIP: start outline for perf logging

tomeichlersmith committed:
outline headers of performance tracker and its measurements

tomeichlersmith committed:
compiling version of separate ROOT dict

attempted to write, not working but I think that's related to a warning
at compilation time from root-cling about "unused class rule"

tomeichlersmith committed:
delete old outline

tomeichlersmith committed:
compiles and can run like normal (no perf logging)

going to want to test perf logging now

tomeichlersmith committed:
need to add header which uses the STL collection manually

tomeichlersmith committed:
functionally logging performance

had to change serialization structure to support uproot-reading of
performance data, just using directories rather than std::map's directly

tomeichlersmith committed:
move filling of by-event data to its own function

hopefully reduces overhead by removing a string comparison

tomeichlersmith committed:
doxygen comments on measurement, add invalidate method

tomeichlersmith committed:
invalidate event measurements on end of each event

tomeichlersmith committed:
update doxy comments of Tracker

EinarElen commented on src/Framework/Process.cxx:
```diff
@@ -105,9 +105,17 @@ Process::Process(const framework::config::Parameters &configuration)
 Process::~Process() {
+  // need to delete the performance object so that it is
+  // written before we close the histogram file below
+  if (performance_) delete performance_;
```
Do we have a reason for not using a unique_ptr?

tomeichlersmith replied:
I tried using a unique_ptr and the issue arises from the `std::move` in `getDummy`. We'd need to define a move constructor to get it working and, since I need to call `reset` anyways to insure that it is destructed at the right time, I opted for a raw pointer.

EinarElen commented on include/Framework/Performance/Tracker.h:
```diff
@@ -0,0 +1,85 @@
+  TTree *event_data_;
+
+  Measurement absolute_start_, absolute_end_;
+  std::map<std::string, Measurement> 
```
Relying on a map and string comparison operations for a performance measurement (at least one without amortization) is slightly scary to me (seems like it is asking for uncertainty/cache invalidation) ((otoh hep simulations are usually slow enough that you can't tell but we should be sure)), but it is a place to start

tomeichlersmith replied:
I don't like it either, but I don't know of a better way to accommodate the different names of the processors.

In a meeting before, I think you mentioned just using indices (referring to the order the processors run in), so I could look at implementing something like that.

EinarElen replied:
I think that would be good to do at some point, but lets try get a workable feature out first! But it could have some non-trivial spooky action at a distance effects

EinarElen commented on src/Framework/Performance/Measurement.cxx:
```diff
@@ -0,0 +1,28 @@
+static long unsigned int get_current_time() {
+  using std::chrono::system_clock;
+  const auto current_time{system_clock::now()};
+  return system_clock::to_time_t(current_time);
```
Any reason for why we are recording time_t? I would preferably have the recordings in the form of std::chrono duration types corresponding to time deltas

tomeichlersmith replied:
The reason is that you'd need to measure time deltas relative to some original time. I don't want to pre-define an original time for anyone since it may change depending on use case.

The biggest example of this is the fact that new branches are allocated and initialized on their first `add`. Perhaps you set your origin time to the 10'th event to avoid this overhead polluting your measurement.

EinarElen replied:
I think what I would find useful is the time delta between two events, i.e. between a begin_X and end_X

tomeichlersmith replied:
Ah that makes sense, how would other measurements fit in? Would we take to measurements of memory/cpu usage for example?

EinarElen replied:
That I'm less sure about but I think runtime and resource usage are quite different kinds of measurements so recording them in the same place is probably going to make things messy/complicated _somewhere_. 

For a memory usage measurement for example, would you really care about the memory use before/after operations? Or would you want some more uniform sampling? Or some average over an operation? 

I'm much more familiar with runtime measurements so I might be wrong!

EinarElen commented on src/Framework/Performance/Tracker.cxx:
```diff
@@ -0,0 +1,171 @@
+  storage_directory_->WriteObject(&absolute_start_, "absolute_start");
+  storage_directory_->WriteObject(&absolute_end_, "absolute_end");
+
+  TDirectory* onProcessStart_d = storage_directory_->mkdir("onProcessStart");
```
This section feels like it could use some DRY

EinarElen commented on src/Framework/Process.cxx:
```diff
@@ -160,9 +175,7 @@ void Process::run() {
-
-    for (auto module : sequence_) module->onFileOpen(outFile);
-
+    onFileOpen(outFile);
```
Having these things factored out is great

EinarElen commented on include/Framework/Performance/Measurement.h:
```diff
@@ -0,0 +1,52 @@
+  /// validity of this measurement (true if `sample` was called, false otherwise)
+  bool valid_{false};
+  /// the time stamp of this measurement (as measured by system clock)
+  long unsigned int time_;
```
See my later comment on why we are recording timestamps as opposed to time deltas

EinarElen reviewed (commented):
This looks great, I'll go ahead and try things out a bit

EinarElen commented:
Either I messed up my config or this broke something 

```shell
/Users/einarelen/ldmx/ldmx-sw/SimCore/src/SimCore/Generators/RootSimFromEcalSP.cxx: In destructor 'virtual simcore::generators::RootSimFromEcalSP::~RootSimFromEcalSP()':
/Users/einarelen/ldmx/ldmx-sw/SimCore/src/SimCore/Generators/RootSimFromEcalSP.cxx:49:11: error: 'class framework::EventFile' has no member named 'close'
   49 |   ifile_->close();
```

EinarElen commented:
Here's a pretty dumb little test config that I used 
```python
from LDMX.Framework import ldmxcfg
p = ldmxcfg.Process('perftest')
p.logPerformance = True
p.maxTriesPerEvent = 1

from LDMX.SimCore import generators as gen
from LDMX.SimCore import simulator
import LDMX.Ecal.EcalGeometry
import LDMX.Ecal.ecal_hardcoded_conditions
import LDMX.Hcal.HcalGeometry
import LDMX.Hcal.hcal_hardcoded_conditions
import LDMX.Ecal.digi as ecal_digi
import LDMX.Ecal.vetos as ecal_vetos
import LDMX.Hcal.digi as hcal_digi
# from Biasing import ecal
mySim = simulator.simulator('perftest')

gun = gen.gun('neutron_gun')
gun.particle='neutron'
# gun.particle='pi0'
# gun.particle='neutron'
gun.position = [0.,0.,240.]
gun.direction = [0.,0.,1.]
gun.energy = 3.0
import sys
det = sys.argv[1]
mySim.setDetector(f'ldmx-det-v{det}')
mySim.beamSpotSmear = [20.,80.,0.]
mySim.description = f'Perf test {det}'
mySim.generators = [gun]
p.sequence = [ mySim,
        hcal_digi.HcalDigiProducer(),
        hcal_digi.HcalRecProducer(),
              ]

p.termLogLevel = 1

##################################################################
# Below should be the same for all sim scenarios

import os
import sys

if 'LDMX_NUM_EVENTS' in os.environ:
    p.maxEvents = int(os.environ['LDMX_NUM_EVENTS'])
else:
    p.maxEvents = int(sys.argv[2])
if 'LDMX_RUN_NUMBER' in os.environ:
    p.run = int(os.environ['LDMX_RUN_NUMBER'])
else:
    p.run = 2

output_base = f'perf_test_{det}'
if len(sys.argv) > 2:
    output_base = sys.argv[3]


p.histogramFile = f'{output_base}_hist.root'
p.outputFiles = [f'{output_base}_events.root']

import LDMX.Ecal.EcalGeometry
import LDMX.Hcal.HcalGeometry


from LDMX.DQM import dqm

p.sequence.extend(
    dqm.hcal_dqm
)
```

My first thought is that for this to be as useful as possible for working with any runtime measurent, we really need to be recording time-deltas and not time points. When trying to find out why something isn't working, you really do want to be able to directly look at a histogram and not series of raw values (or we can add something that does this somewhere else but it does need to be tightly integrated)

tomeichlersmith commented:
https://github.com/LDMX-Software/Framework/commit/f3036b1eff780875e57492ee7e73236b1549a0f8 is why you see the compilation error with `RootSimFromEcalSP` . Not sure if we want to put a `EventFile::close` back in or if we are ok with having downstream patches necessary for this minor version.

(I'm personally ok with forcing downstream patches for minor version bumps.)

EinarElen commented:
> [f3036b1](https://github.com/LDMX-Software/Framework/commit/f3036b1eff780875e57492ee7e73236b1549a0f8) is why you see the compilation error with `RootSimFromEcalSP` . Not sure if we want to put a `EventFile::close` back in or if we are ok with having downstream patches necessary for this minor version.
> 
> (I'm personally ok with forcing downstream patches for minor version bumps.)

Yeah that's probably going to be fine as long as it is referenced somewhere, it isn't changing the format of any data

tomeichlersmith committed:
refactor measurement to focus on duration of time

EinarElen commented:
I was literally just about to write a suggestion for roughly what you just implemented lol, not bad

EinarElen commented:
When testing 5789b4e4bb3986565fb0aa65b55b8ee51fc9c8eb with the config above, I get no entries in the perf tree for the simulator. Not sure why? 

I'm also consistently getting two trees called by_event

Update: The simulator is in there, just forgot that it would be called just "perftest" in this config.

EinarElen commented on src/Framework/Performance/Measurement.cxx:
```diff
@@ -12,13 +12,17 @@ static long unsigned int get_current_time() {
+void Measurement::end() {
   valid_ = true;
-  time_ = get_current_time();
+  duration_ = get_current_time() - start_;
```
I'm not 100% sure this will do anything sensible, since start_/get_current_time are std::time_t objects. I don't think subtracting them gives a valid duration. 

I would make start_ a std::chrono time_point and have some using declarations in the header that determine 
- What clock to use (e.g. `using clock = std::chrono::system_clock` `using time_point = clock::time_point`
- What duration to save (potentially something you could customize later?)
`using duration = std::chrono::nanoseconds` 

Then duration_ woudl be recorded as 
```
std::chrono::duration_cast<duration>(get_current_time() - start_).count()
```

or something like this

tomeichlersmith replied:
Yea, I'm doing another rewrite where I use chrono duration so I can get the seconds as a `double`.

EinarElen reviewed (commented):
I think this approach both made the setup easier to use and to read. Here's a quick example measurement where I changed the duration as I suggested above

<img width="1165" alt="Screenshot 2023-12-01 at 18 03 20" src="https://github.com/LDMX-Software/Framework/assets/14233914/5e033632-ca8e-4b32-a393-14e98ff2538e">

EinarElen commented:
Here's a first interesting result:

This is simulation time for 1000 neutrons fired into the hcal with v12 geometry
<img width="1100" alt="Screenshot 2023-12-01 at 18 06 19" src="https://github.com/LDMX-Software/Framework/assets/14233914/494e854c-c73d-4967-b250-d4df7e029cde">
and the same for v14
<img width="1165" alt="Screenshot 2023-12-01 at 18 03 20" src="https://github.com/LDMX-Software/Framework/assets/14233914/30fe14c2-d7ab-448b-86c7-feb0a09c4079">

Seems like the v14 geometry changes did not slow down the simulation!

tomeichlersmith committed:
refactor to focus on using std::chrono for timing

- more precise and can then extract a double representing seconds elapsed
which is a way more understandable unit to come out of looking at the
resulting file
- abandoned the measurement class customizing serialization and just
  writing single-element arrays instead, but what can you do
https://root-forum.cern.ch/t/writing-simple-variable-in-root-files/11094

tomeichlersmith committed:
re-enable custom class serialization for performance logging

I learned how to mark class members as transient [0] which made me
realize that having a special Timer class will make the code easier to
read and the data easier to interpret, I'm happier with this design than
previous designs so I spent a bit more effort in documentation.

[0] https://root.cern/manual/io/#excluding-data-members-from-io

tomeichlersmith commented:
Alright, I'm marking this as ready-for-review since I have a solution that I actually like. It uses std::chrono under-the-hood for timing but then casts those durations to specific units for serialization. In https://github.com/LDMX-Software/framework-testbench, I've added a config that helps test this performance logging machinery `config/log-performance.py` and a quick analysis script to print the results to check them.

This shows that the timers are being correctly invalidated between events (`-1` is used for start_time and duration as the invalid value) and the processors are being faithfully "registered" with the performance tracking system.

```
tom@framework-testbench:~$ fire config/log-performance.py                                                                                                                                                                                     
---- LDMXSW: Loading configuration --------                                                                                                                                                                                                   
---- LDMXSW: Configuration load complete  --------                                                                                                                                                                                            
---- LDMXSW: Starting event processing --------                                                                                                                                                                                               
 [ Process ] 1 : Processing 1 Run 1 Event 1  (2023-12-04 16:15:09.463132000+0000)                                                                                                                                                             
 [ Process ] 1 : Processing 2 Run 1 Event 2  (2023-12-04 16:15:09.478684000+0000)                                                                                                                                                             
 [ Process ] 1 : Processing 2 Run 1 Event 2  (2023-12-04 16:15:09.478826001+0000)                                                                                                                                                             
 [ Process ] 1 : Processing 3 Run 1 Event 3  (2023-12-04 16:15:09.478987000+0000)                                                                                                                                                             
 [ Process ] 1 : RunHeader { run: 1, numTries: 4, detectorName: , description:                                                                                                                                                                
  intParameters:                                                                                                                                                                                                                              
    RandomNumberMasterSeed[prod] = 1                                                                                                                                                                                                          
  floatParameters:                                                                                                                                                                                                                            
  stringParameters:                                                                                                                                                                                                                           
}                                                                                                                                                                                                                                             
---- LDMXSW: Event processing complete  --------   
tom@framework-testbench:~$ python3 ana/print-performance.py test_performance.root 
absolute
0.080721518 1701706509462766424
onProcessStart
   __ALL__;1  1.649e-05 1701706509462916012
   MultiTry;1 2.487e-06 1701706509462924301
   Recon;1    2.100e-07 1701706509462932052
onProcessEnd
   __ALL__;1  2.866e-06 1701706509543449644
   MultiTry;1 5.670e-07 1701706509543450937
   Recon;1    2.180e-07 1701706509543452159
onFileOpen
   __ALL__;1  5.616e-06 1701706509463029186
   MultiTry;1 3.830e-07 1701706509463033368
   Recon;1    9.300e-08 1701706509463034574
onFileClose
   __ALL__;1  9.240e-07 1701706509479079779
   MultiTry;1 2.260e-07 1701706509479080088
   Recon;1    8.000e-08 1701706509479080499
beforeNewRun
   __ALL__;1  2.343e-06 1701706509463087350
   MultiTry;1 2.080e-07 1701706509463088821
   Recon;1    1.060e-07 1701706509463089479
onNewRun
   __ALL__;1  3.595e-05 1701706509463089785
   MultiTry;1 1.350e-07 1701706509463125263
   Recon;1    7.700e-08 1701706509463125540
{
  "completed": "array([ True, False,  True,  True])",
  "__ALL__": {
    "start_time_": "array([1701706509463378422, 1701706509478749743, 1701706509478853900,\n       1701706509479016118])",
    "duration_": "array([8.05879e-04, 5.12100e-05, 9.13860e-05, 4.22580e-05])"
  },
  "MultiTry": {
    "start_time_": "array([1701706509463379031, 1701706509478751018, 1701706509478854315,\n       1701706509479016507])",
    "duration_": "array([ 6.28204e-04, -1.00000e+00,  8.18650e-05,  3.58290e-05])"
  },
  "Recon": {
    "start_time_": "array([1701706509464008068,                  -1, 1701706509478936706,\n       1701706509479052849])",
    "duration_": "array([ 1.75532e-04, -1.00000e+00,  8.43000e-06,  5.38800e-06])"
  }
}
```

tomeichlersmith commented on src/Framework/Process.cxx:
```diff
@@ -411,18 +443,55 @@ bool Process::process(int n, Event& event) const {
+      if (performance_) performance_->end(performance::Callback::process, i_proc);
     }
   } catch (AbortEventException &) {
+    if (performance_) {
+      performance_->end(performance::Callback::process, 0);
+      performance_->end_event(false);
+    }
```
Do we want to end the process callback for the processor that threw `AbortEventException`? Right now it doesn't (which I think makes sense), but I could forsee someone wanting to know how quick aborted events take.

EinarElen replied:
Do we get an `AbortEventException` when we abort at G4 event (e.g. due to a filter)? It _might_ be useful to have some idea of what time is spent in rejected vs accepted events. I can't think of any other places it would be useful

tomeichlersmith replied:
Yes, we do get an `AbortEventException` from our Simulator when we abort a G4 event.
(They are thrown with the `abortEvent` protected member function of processors.)

I will update the code to `end` the timer for the processor that threw `AbortEventException`.

tomeichlersmith committed:
add more docs on Callback and cleanup

tomeichlersmith commented on include/Framework/Performance/Tracker.h:
```diff
@@ -0,0 +1,68 @@
+  void absolute_start();
+  /// literally last line of Process::run (if run compeletes without error)
+  void absolute_end();
+  /// start the timer for a specific callback and specific processor
+  void start(Callback cb, std::size_t i_proc);
+  /// end the timer for a specific callback and specific processor
+  void end(Callback cb, std::size_t i_proc);
```
I'm tempted to move the run-time `Callback` argument to a compile-time argument since these are only ever used at compile time.

I also don't think this would have much of an effect though.

tomeichlersmith replied:
I've made the executive decision to be lazy :monocle_face: 

I don't think it will have any effect that is distinguishable even if we have optimizations off. If we have optimizations on, I'm fairly certain the compiler will be able to deduce the same assembly without requiring us to do fancy footwork with `template` in order to move `cb` to a compile-time argument.

EinarElen replied:
This is resolved but we could always try sprinkling some constexpr here and there (templating seems like a lot of effort)

tomeichlersmith replied:
I made the `to_index` function `constexpr` but I'm not really sure where to go from here - feel free to try something yourself if you have an idea? I'm not as well-read on `constexpr` as you I believe.

EinarElen commented on include/Framework/Performance/Timer.h:
```diff
@@ -0,0 +1,81 @@
+ * into seconds (including sub-second increments). Since ROOT doesn't have a
+ * dictionary for serializing these std::chrono classes (and I'm not interested
+ * in making one), I simply mark those helper-members as "transient" so they
+ * are ignored by the dictionary generation. The atomic members start_time_ and
```
Very minor comment but atomic has a particular technical definition in the context of C++ so maybe we'd want to call it something else? Or do we mean atomic in the technical sense?

tomeichlersmith replied:
I guess I don't mean atomic in the [std::atomic](https://en.cppreference.com/w/cpp/atomic/atomic), I'll change the word to fundamental like https://en.cppreference.com/w/cpp/language/types

EinarElen commented on include/Framework/Performance/Timer.h:
```diff
@@ -0,0 +1,81 @@
+ */
+class Timer {
+  using clock = std::chrono::high_resolution_clock;
+  std::chrono::time_point<clock> start_; //! not serialized, just for measurement purposes
```
Might be worth highlighting once more to be extra clear that the //! syntax is what determines that it isn't serialized with a comment above

EinarElen commented on include/Framework/Performance/Timer.h:
```diff
@@ -0,0 +1,81 @@
+   */
+  long int start_time_{-1};
+  /**
+   * duration timer was run for in seconds (with sub-second increments included)
```
?

tomeichlersmith replied:
I'm trying to point out that the unit is seconds but that milliseconds (and other sub-second intervals) are included in the `double` so users can still do sub-second measurements with this data.

EinarElen commented on include/Framework/Performance/Timer.h:
```diff
@@ -0,0 +1,81 @@
+  /// start the timer
+  void start();
+  /// end the timer
+  void end();
```
Why not stop?

tomeichlersmith replied:
No reason - I'll update it since `stop` is more clear.

EinarElen commented on include/Framework/Performance/Callback.h:
```diff
@@ -0,0 +1,38 @@
+/**
+ * Conver the Callback enum into a human-readable name
+ */
+const char* to_name(Callback c);
```
Can we do this with a std::string instead of C-style strings? I'm guessing that will add some annoying .c_str() because... ROOT? Or is the to_name function used in any critical paths?

tomeichlersmith replied:
Yes, the only reason I am using C-style strings is because ROOT - currently, the only place this function is used is when the directory for a callback is made in the ROOT file we are writing to.

No critical paths, this is just called once per callback at writing-time.

EinarElen commented on src/Framework/Performance/Callback.cxx:
```diff
@@ -0,0 +1,38 @@
+      return "process";
+      break;
+    default:
+      return "WTF";
```
Raise exception? Or don't have a default at all so that missing a case statement will be a compiler warning. 

Or combine the two, skip the default and throw if we ever exit the switch statement (you know in case someone ignores a warning...)

tomeichlersmith replied:
Good point, forward thinking, I was just throwing stuff together. Will update with more future-proofed implementation.

EinarElen commented on src/Framework/Performance/Tracker.cxx:
```diff
@@ -0,0 +1,109 @@
+   * TTree as branches
+   *
+   * @note This is where we connect the _address_ of our timer
+   * to the TTree for serialization. This effectively means that
```
Highlight this in the header as well, by the corresponding variable

tomeichlersmith committed:
end the timer for the processor that threw the AbortEventException

tomeichlersmith committed:
more thorough documentation of transient member variables

tomeichlersmith committed:
refactor to_name(Callback) with an eye towards future devs

- return a std::string so its more modern
- throw an exception if the function isn't left within the switch
  statement
- leave out the `default:` clause so there is a compiler warning if an
  unmatched Callback is given to to_name

tomeichlersmith committed:
add comment on member variable about how its addresses should not change

tomeichlersmith committed:
refactor naming related to timers end->stop

store the time-points as begin and end (since they are nouns) and have
the functions that are actions (i.e. verbs) be named start/stop. We were
already using start as the function name so I just needed to change the
end function names to stop

tomeichlersmith committed:
make the to_index(Callback) function constexpr

hopefully this helps C++ inline it when compiling the downstream calls

tomeichlersmith committed:
add advice on how to access the perf TTree data with uproot

EinarElen reviewed (approved):
This looks really great! 

Can I ask for a corresponding note about the feature in the regular docs? Would be good to have a place to point people too (other than the code/doxygen)

tomeichlersmith commented:
That makes sense :+1: Please help me remember - I plan to go through the "basic-running" docs when I am merging the updated Framework into ldmx-sw.
